allennlp/models/bert_for_classification.py

```diff
@@ -0,0 +1,153 @@
+
+from allennlp.data.vocabulary import Vocabulary
+from allennlp.models.model import Model
+from allennlp.modules.token_embedders.bert_token_embedder import PretrainedBertModel
+from allennlp.nn.initializers import InitializerApplicator
+from allennlp.training.metrics import CategoricalAccuracy
+
+
+@Model.register("bert_for_classification")
+class BertForClassification(Model):
+    """
+    An AllenNLP Model that runs pretrained BERT,
+    takes the pooled output, and adds a Linear layer on top.
+    If you want an easy way to use BERT for classification, this is it.
+    Note that this is a somewhat non-AllenNLP-ish model architecture,
+    in that it essentially requires you to use the "bert-pretrained"
+    token indexer, rather than configuring whatever indexing scheme you like.
+
+    See `allennlp/tests/fixtures/bert/bert_for_classification.jsonnet`
+    for an example of what your config might look like.
+
+    Parameters
+    ----------
+    vocab : ``Vocabulary``
+    bert_model : ``Union[str, BertModel]``
+        The BERT model to be wrapped. If a string is provided, we will call
+        ``BertModel.from_pretrained(bert_model)`` and use the result.
+    num_labels : ``int``, optional (default: None)
+        How many output classes to predict. If not provided, we'll use the
+        vocab_size for the ``label_namespace``.
+    index : ``str``, optional (default: "bert")
+        The index of the token indexer that generates the BERT indices.
+    label_namespace : ``str``, optional (default : "labels")
+        Used to determine the number of classes if ``num_labels`` is not supplied.
+    trainable : ``bool``, optional (default : True)
+        If True, the weights of the pretrained BERT model will be updated during training.
+        Otherwise, they will be frozen and only the final linear layer will be trained.
+    initializer : ``InitializerApplicator``, optional
+        If provided, will be used to initialize the final linear layer *only*.
+    """
+    def __init__(self,
+                 vocab: Vocabulary,
+                 bert_model: Union[str, BertModel],
+                 dropout: float = None,
+                 num_labels: int = None,
+                 index: str = "bert",
+                 label_namespace: str = "labels",
+                 trainable: bool = True,
+                 initializer: InitializerApplicator = InitializerApplicator()) -> None:
+        super().__init__(vocab)
+
+        if isinstance(bert_model, str):
+            self.bert_model = PretrainedBertModel.load(bert_model)
+        else:
+            self.bert_model = bert_model
+
+        self.bert_model.requires_grad = trainable
+
+        in_features = self.bert_model.config.hidden_size
+
+        if num_labels:
+            out_features = num_labels
+        else:
+            out_features = vocab.get_vocab_size(label_namespace)
+
+        if dropout is not None:
+            self._dropout = torch.nn.Dropout(p=dropout)
```

Comment:
I'm pretty sure we don't need to do this pattern anymore; I'm not sure why we did it originally, but if you set the default dropout value to `0.0` instead of `None`, you don't need these conditionals (here or in `forward`), and @yizhongw found there was no memory or speed hit for doing so.